### PR TITLE
Rework text on exceptions

### PR DESCRIPTION
**Pull request by encukou**
This ias a major update on the text on Exceptions.
Also, coverage of *nested tracebacks* is split out and moved next to *defining functions*.

Please review the changes as a whole, don't look at individual commits. But, to make cherry-picking easier, don't squash when merging.

### CODE REVIEW

**Comment by kokes** [comment]
Not sure if fits within the scope of this lesson, but I would maybe mention that exception-driven development can be oftentimes avoided and doing so can lead to more legible and understandable code.

Things like
```python
try:
  foo = bar['bak']
except KeyError:
  ...
```
instead of `foo.get` are all too common and I've even seen bugs introduced this way (e.g. by including more things in the try block). You somewhat alluded to this in your "we can check if the string is parseable to an int, but it's best to run `int` directly" - but this is not always the case and we can prevent try/except blocks in many cases.

**Comment by encukou** [comment]
That particular example won't work because this lesson is before `dict`.
I can add some text on keeping the `try` block minimal. But I'd prefer to do it in a separate PR, this one is already quite large.

Also, I'm afraid mentioning the problem won't mean you won't see mistakes like that. It's something you learn by practice.

File: lessons/beginners/exceptions/index.md [hunk 1]
```diff
@@ -1,70 +1,98 @@
...
-    return 1 / delitel
-
-print(vnejsi_funkce())
+def nacti_cislo():
+    odpoved = input('Zadej číslo: ')
+    return int(odpoved)
 ```
 
-<!-- XXX: Highlight the line numbers -->
+Když uživatel nezadá číslice, ale třeba text `cokolada`,
```
Comment (fivaldi) [inline lessons/beginners/exceptions/index.md:14]: `uživatel(ka)`

File: lessons/beginners/exceptions/index.md [hunk 2]
```diff
@@ -74,86 +102,57 @@ BaseException
...
-        try:
-            return int(odpoved)
-        except ValueError:
-            print('To nebylo číslo! Zkus to znovu.')
-```
+Co ale dělat, kdyš uživatel chce ukončit program a zmáčkne
+<kbd>Ctrl</kbd>+<kbd>C</kbd>?
+Nebo když se mu porouchá klávesnice a selže funkce `input`?
+Nejlepší reakce na takovou nečekanou situaci ukončit program a informovat
+uživatele (nebo lépe, programátora), že (a kde) je něco špatně.
```
Comment (fivaldi) [inline lessons/beginners/exceptions/index.md:148]: Reword: `uživatele, že je něco špatně, a bokem i programátor(k)y, kde a kdy k tomu došlo.` or so...

File: lessons/beginners/exceptions/index.md [hunk 3]
```diff
@@ -182,54 +181,49 @@ finally:
...
-    except ValueError:
-        print('To nebylo číslo!')
-    else:
-        if strana <= 0:
-            print('To nedává smysl!')
-        else:
-            break
+* Zadal-li ono `-5` uživatel, je potřeba mu vynadat a zeptat se znovu.
+* Naměřil-li `-5` nějaký robotický aparát, je potřeba ho líp zkalibrovat.
+* Vyšel-li čtverec se stranou `-5` v nějakém výpočtu, je nejspíš potřeba opravit
```
Comment (fivaldi) [inline lessons/beginners/exceptions/index.md:195]: Reword without `-li`, masculine and (too strong) `vynadat` -> rather `upozornit`.

File: lessons/beginners/exceptions/index.md [hunk 4]
```diff
@@ -182,54 +181,49 @@ finally:
...
+Samotná funkce `obsah_ctverce` ale „neví“, proč ji někdo volá.
+Jejím úkolem je jen něco spočítat.
+Měla by být použitelná ve všech případech výše – a v mnoha dalších.
 
-```
+Když někdo zavolá `obsah_ctverce(-5)`, *neexistuje* správný výsledek, který by
+funkce mohla vrátit.
+Místo vrácení výsledku musí tato funkce *signalizovat chybu*.
+S tou se pak může program, který `obsah_ctverce(-5)` zavolal,
+vypořádat – vynadat uživateli, zkalibrovat měřák, nebo, pokud na chybu není
```
Comment (fivaldi) [inline lessons/beginners/exceptions/index.md:206]: Again: `vynadat`

File: lessons/beginners/exceptions/index.md [hunk 5]
```diff
@@ -182,54 +181,49 @@ finally:
...
+připravený, sám skončit s chybou (a upozornit tak programátora, že je něco
```
Comment (fivaldi) [inline lessons/beginners/exceptions/index.md:207]: programátora/-ku

File: lessons/beginners/nested-traceback/index.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  Měla by funkce `podel` zjistit, jestli dostane nulu, a nějak na to
+  zareagovat?
+- Na řádku 8, ve funkci `podel_nulou`, kde voláš `podel` s dělitelem 0.
+  Má to tak opravdu být?
+- Na řádku 13, ve funkci `ukaz_priklad`, kde voláš funkci `podel_nulou`.
+  Kdybys to nedělal{{a}}, chyba by taky nevznikla!
+- Na řádku 16, ne ve funkci, když voláš funkci `ukaz_priklad`.
+  Možná by to chtělo použít jiný příklad?
+
+Python (a asi ani ty) nemůže vědět, co tím programem programátor myslel,
```
Comment (fivaldi) [inline lessons/beginners/nested-traceback/index.md:61]: autor(ka) programu myslel(a)

File: lessons/beginners/nested-traceback/index.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+a kdy by tedy bylo nejlepší chybu opravit.
+Ukáže tedy v programu všechna místa, která k chybě vedla.
+Je na tobě abys z nich vybral{{a}} to nejvhodnější a zaměřil{{a}} se na něj.
+
+Tahle ukázka je samozřejmě jen teoretická, ale v reálných programech vypadá
+hlášení chyb stejně.
+Až takovou složitou hlášku uvidíš, nepanikař!
+Python se snaží co nejvíc napovědět a usnadnit ti chybu najít a opravit.
+Sice stroze a anglicky, ale snaží.
+Vyjdi mu vstříc a nauč tyhle se hlášky číst.
```
Comment (fivaldi) [inline lessons/beginners/nested-traceback/index.md:71]: Word order.

**Comment by fivaldi** [comment]
Please see my comments. With this PR, the materials won't grow too much ( +209 −139 ), but the section ` Chybové hlášky ze zanořených funkcí` should be somehow marked as "advanced"/"additional" for the live course. Or, subject for additional home-learning or for recap lessons.

**Comment by fivaldi** [comment]
> That particular example won't work because this lesson is before `dict`.
> I can add some text on keeping the `try` block minimal. But I'd prefer to do it in a separate PR, this one is already quite large.
> 
> Also, I'm afraid mentioning the problem won't mean you won't see mistakes like that. It's something you learn by practice.

Maybe give some simple general advice(s) like:

1. Use try-except blocks (always with caution(!)) for exceptions which you expect. (_This is already mentioned in the materials._)

1. They're here exactly for this/your purpose! But keep in mind that some functions/methods already provide arguments (generally _mechanisms_) for providing defaults instead of missing/invalid values, so you don't even need to take care of exception handling.

File: lessons/beginners/nested-traceback/index.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+def ukaz_priklad():
+    """Spočítá ukázkový příklad a výsledek vypíše (nevrátí!)"""
+    vysledek = podel_nulou(3)       # řádek 13
+    print(vysledek)
+
+ukaz_priklad()                      # řádek 16
+```
+
+Co se stane, ktyž tohle pustíš?
```
Comment (befeleme) [inline lessons/beginners/nested-traceback/index.md:32]:
```suggestion
Co se stane, když tohle pustíš?
```

File: lessons/beginners/exceptions/index.md [hunk 9]
```diff
@@ -1,70 +1,98 @@
...
+Kde ale vzít funkci `obsahuje_jen_cislice`?
+Nemá smysl ji psát znovu – funkce `int` sama nejlíp pozná, co se dá převést na
+číslo a co ne.
+A dokonce nám to dá vědět – chybou, kterou můžeš *zachytit*.
+
+> [note]
+> Ono „obsahuje_jen_cislice“ v Pythonu existuje. Dokonce několikrát.
+> Místo řešení problému to ale spíš ilustruje, v čem problém spočívá:
+> * Řetězcová metoda `isnumeric` vrací `True` pokud řetězec obsahuje číslice:
+>   `'123'.isnumeric()` je pravda; ``'abc'.isnumeric()` nepravda.
```
Comment (befeleme) [inline lessons/beginners/exceptions/index.md:56]:
```suggestion
>   `'123'.isnumeric()` je pravda; `'abc'.isnumeric()` nepravda.
```

File: lessons/beginners/nested-traceback/index.md (hunk 7 above)
Comment (encukou) [reply lessons/beginners/nested-traceback/index.md:71]:
```suggestion
Vyjdi mu vstříc a nauč se tyhle hlášky číst.
```

File: lessons/beginners/exceptions/index.md (hunk 3 above)
Comment (encukou) [reply lessons/beginners/exceptions/index.md:195]: I don't think `vynadat` is too strong here. I've used it quite successfully to refer to "bad user input" messages.

Comment (encukou) [reply lessons/beginners/exceptions/index.md:195]: Also about `-li`: I don't understand why that needs reqording.

**Comment by encukou** [comment]
> the section Chybové hlášky ze zanořených funkcí should be somehow marked as "advanced"/"additional" for the live course. Or, subject for additional home-learning or for recap lessons.

Yes, this does work pretty well as a home learning assignment, and when it's split away from the main lesson, it's easy to mark it that way in the live courses.
But there's (so far) no way to mark it as a home assignment for live courses from the self-study materials.

> Maybe give some simple general advice(s) / best practices like:
> 
>     1. Use try-except blocks (always with caution) for exceptions which you expect to happen. (_This is already mentioned in the materials._)
> 
>     2. They're here exactly for this/your purpose! But keep in mind that some functions/methods already provide arguments (generally _mechanisms_) for providing defaults instead of missing/invalid values, so you don't even need to take care of exception handling.
> 
>     3. Use try-except blocks with double caution(!) for unknown/general exceptions **only** in places where you want your programme to be bullet-proof. (_Rare situations._)

I'm skeptical that a numbered list will work, at least without examples, much more discussion and/or assignments. Specifically, 2. will generally not come into play before students get to `dict.get`, and 3. until they need to write something like an event loop, REPL or request handling server. By that time, they'll probably forget the advice.
So, IMO, mentioning 2. when `dict.get` is introduced, instead of here, would be more effective. And 3. is out of scope of this course.

What I will be working on in future courses is explaining that the `try` block should be minimal in size (and why), in addition to `except` limiting the types of handled exceptions.

**Comment by fivaldi** [comment]
Thanks for fixing typos and leaving comments on reviews. I guess it can be merged. Please someone with merge privs to do so.

+1

**Comment by encukou** [comment]
Thanks for the review!

**Pull request merged**